astropy/io/ascii/ecsv.py

```diff
@@ -175,13 +180,13 @@ def get_cols(self, lines):
                     setattr(col, attr, header_cols[col.name][attr])
 
             col.dtype = header_cols[col.name]['datatype']
-            # Require col dtype to be a valid ECSV datatype. However, older versions
-            # of astropy writing ECSV version 0.9 and earlier had inadvertently allowed
-            # numpy datatypes like datetime64 or object or python str, which are not in the ECSV standard.
-            # For back-compatibility with those existing older files, allow reading with no error.
-            if col.dtype not in ECSV_DATATYPES and self.ecsv_version > (0, 9, 0):
-                raise ValueError(f'datatype {col.dtype!r} of column {col.name!r} '
-                                 f'is not in allowed values {ECSV_DATATYPES}')
+            # Warn if col dtype is not a valid ECSV datatype, but allow reading for
+            # back-compatibility with existing older files that have numpy datatypes
+            # like datetime64 or object or python str, which are not in the ECSV standard.
+            if col.dtype not in ECSV_DATATYPES:
+                msg = (f'unexpected datatype {col.dtype!r} of column {col.name!r} '
+                       f'is not in default datatypes {ECSV_DATATYPES}, using anyway.')
```

Comment:
The "unexpected results" mostly applies to the `object` type but I think it's a good idea to call out that they need to be careful.
```suggestion
                       f'is not in allowed ECSV datatypes {ECSV_DATATYPES}. '
                       'Using anyway as a numpy dtype but beware since unexpected results '
                       'are possible.)
```